### PR TITLE
load active delegates from stateDB into candidate center upon protocol.Start()

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: action/protocol/staking/protocol.go [hunk 1]
```diff
@@ -109,3 +121,33 @@ func (p *Protocol) Register(r *protocol.Registry) error {
...
+	candPkHash, candValue, err := stateDB.Filter(factory.CandidateNameSpace, func(k, v []byte) bool {
+		return bytes.HasPrefix(k, []byte{candidate})
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(candValue) == 0 {
+		return nil
+	}
```
Comment (yutongp) [inline action/protocol/staking/protocol.go]: can be removed. len == 0 will just pass below for loop

File: action/protocol/staking/protocol.go [hunk 2]
```diff
@@ -109,3 +121,33 @@ func (p *Protocol) Register(r *protocol.Registry) error {
...
+
+	// decode the candidate and put into candidate center
+	for i := range candPkHash {
+		c := &Candidate{}
+		if err := c.Deserialize(candValue[i]); err != nil {
+			return errors.Wrapf(err, "failed to deserialize candidate bytes %v", candValue[i])
+		}
+		p.inMemCandidates.Put(c)
```
Comment (yutongp) [inline action/protocol/staking/protocol.go]: Put return error

File: action/protocol/staking/protocol.go (hunk 1 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go]: updated in latest push

File: action/protocol/staking/protocol.go (hunk 2 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go]: updated in latest push

File: action/protocol/staking/protocol.go [hunk 3]
```diff
@@ -109,3 +121,31 @@ func (p *Protocol) Register(r *protocol.Registry) error {
...
+	for i := range candPkHash {
+		c := &Candidate{}
+		if err := c.Deserialize(candValue[i]); err != nil {
+			return errors.Wrapf(err, "failed to deserialize candidate bytes %v", candValue[i])
+		}
+		if err := p.inMemCandidates.Put(c); err != nil {
+			return err
+		}
+	}
+	return nil
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: this is supposed to be in Start

File: action/protocol/staking/protocol.go [hunk 4]
```diff
@@ -109,3 +121,31 @@ func (p *Protocol) Register(r *protocol.Registry) error {
...
+func (p *Protocol) CreateGenesisStates(ctx context.Context, sm protocol.StateManager) error {
+	if p.inMemCandidates.Size() > 0 {
+		return nil
+	}
+
+	// read all candidates from stateDB
+	stateDB := sm.GetDB()
+	candPkHash, candValue, err := stateDB.Filter(factory.CandidateNameSpace, func(k, v []byte) bool {
+		return bytes.HasPrefix(k, []byte{candidate})
+	})
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: Let me think a way to not add back getdb

Comment (dustinxie) [reply action/protocol/staking/protocol.go]: stateDB contains an underlying KVStore, i think it is OK to expose that for use

suggest we not spend time/effort on finding a way to not add back getdb

File: action/protocol/staking/protocol.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go]: yes, but Start() does not have state manager

File: action/protocol/staking/protocol.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ import (
 	"github.com/iotexproject/iotex-core/action"
 	"github.com/iotexproject/iotex-core/action/protocol"
 	"github.com/iotexproject/iotex-core/pkg/log"
+	"github.com/iotexproject/iotex-core/state/factory"
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: we may need to move CandidateNamespace to this protocol as it is only used here.

**Pull request merged**